File: mobile/android/app/src/main/kotlin/com/hearth/ui/FamilyScreen.kt
package com.hearth.ui

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.LargeTopAppBar
import androidx.compose.material3.ListItem
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.nestedscroll.nestedScroll
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.hearth.data.FamilyCall
import com.hearth.data.Person
import com.hearth.data.Relationship
import com.hearth.data.Repo
import com.hearth.data.Sex
import com.hearth.family.Origin
import com.hearth.family.ParentCall
import com.hearth.family.alleleOrigins
import com.hearth.family.layoutPedigree
import com.hearth.kb.Kb
import com.hearth.kb.KbEntry
import com.hearth.kb.searchKb
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private val RSID = Regex("^rs\\d+$", RegexOption.IGNORE_CASE)

/**
 * Family lookup (FamilyPage.tsx): find a marker by rsid, gene, drug or condition, then see every
 * person's genotype at it and, when parents are set, who inherited which allele from whom.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FamilyScreen(repo: Repo, kb: Kb) {
    val t = LocalStrings.current
    val scope = rememberCoroutineScope()
    var q by rememberSaveable { mutableStateOf("") }
    var looked by rememberSaveable { mutableStateOf<String?>(null) }
    var rows by remember { mutableStateOf<List<FamilyCall>?>(null) }
    var persons by remember { mutableStateOf(emptyList<Person>()) }
    var relationships by remember { mutableStateOf(emptyList<Relationship>()) }
    fun lookup(rsid: String) = scope.launch {
        q = rsid
        looked = rsid
        rows = withContext(Dispatchers.IO) { repo.familyAt(rsid) }
    }
    LaunchedEffect(Unit) {
        withContext(Dispatchers.IO) { repo.listPersons() to repo.listRelationships() }.let { (p, r) -> persons = p; relationships = r }
        looked?.let { lookup(it) }
    }
    val isRsid = RSID.matches(q.trim())
    val hits = if (isRsid) emptyList() else searchKb(kb, q).take(12)
    val entry = looked?.let { l -> kb.entries.firstOrNull { it.rsid == l } }

    val topBar = TopAppBarDefaults.exitUntilCollapsedScrollBehavior()
    Scaffold(
        modifier = Modifier.nestedScroll(topBar.nestedScrollConnection),
        topBar = {
            LargeTopAppBar(
                title = { Text(t("familyPage.title")) },
                scrollBehavior = topBar,
                colors = TopAppBarDefaults.largeTopAppBarColors(
                    containerColor = MaterialTheme.colorScheme.background,
                    scrolledContainerColor = MaterialTheme.colorScheme.surfaceContainer,
                ),
            )
        },
        containerColor = MaterialTheme.colorScheme.background,
    ) { padding ->
        LazyColumn(Modifier.fillMaxSize(), contentPadding = PaddingValues(top = padding.calculateTopPadding(), bottom = 24.dp)) {
            item(key = "search") {
                Row(Modifier.padding(horizontal = 16.dp, vertical = 8.dp), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    OutlinedTextField(
                        value = q,
                        onValueChange = { q = it },
                        placeholder = { Text(t("familyPage.searchPlaceholder")) },
                        leadingIcon = { Icon(Icons.Filled.Search, contentDescription = null) },
                        singleLine = true,
                        keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.None, autoCorrectEnabled = false, imeAction = ImeAction.Search),
                        keyboardActions = KeyboardActions(onSearch = { if (isRsid) lookup(q.trim()) }),
                        modifier = Modifier.weight(1f),
                    )
                    Button(enabled = isRsid, onClick = { lookup(q.trim()) }) { Text(t("familyPage.lookUp")) }
                }
            }
            items(hits, key = { "hit-${it.rsid}" }) { e ->
                ListItem(
                    headlineContent = { Text("${e.gene} — ${e.name}") },
                    overlineContent = { Text(e.rsid) },
                    supportingContent = { Text(e.summary, maxLines = 2) },
                    modifier = Modifier.clickable { lookup(e.rsid) },
                )
            }
            val found = rows
            if (found != null && looked != null) {
                item(key = "head") {
                    Column(Modifier.padding(16.dp)) {
                        Text(looked!!, style = MaterialTheme.typography.headlineSmall)
                        if (entry != null) Text("${entry.gene} — ${entry.name}", color = MaterialTheme.colorScheme.onSurfaceVariant)
                    }
                }
                if (found.isNotEmpty() && relationships.isNotEmpty()) {
                    item(key = "tree") {
                        Column(Modifier.padding(horizontal = 16.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                            Text(t("familyPage.whoInherited"), style = MaterialTheme.typography.titleMedium)
                            InheritanceTree(persons, relationships, found, entry)
                        }
                    }
                }
                if (found.isEmpty()) {
                    item(key = "none") { Text(t("familyPage.notGenotyped"), Modifier.padding(16.dp), color = MaterialTheme.colorScheme.onSurfaceVariant) }
                } else {
                    item(key = "tableHead") {
                        GenotypeRow(t("familyPage.person"), t("familyPage.chr"), t("familyPage.position"), t("familyPage.genotype"), header = true)
                        HorizontalDivider(Modifier.padding(horizontal = 16.dp))
                    }
                    items(persons, key = { "row-${it.id}" }) { p ->
                        val r = found.firstOrNull { it.personId == p.id }?.call
                        GenotypeRow(p.displayName, r?.chromosome ?: "–", r?.position?.toString() ?: "–", if (r != null) "${r.a1}/${r.a2}" else t("familyPage.notOnChip"), muted = r == null)
                    }
                }
            }
        }
    }
}

@Composable
private fun GenotypeRow(person: String, chr: String, position: String, genotype: String, header: Boolean = false, muted: Boolean = false) {
    val style = if (header) MaterialTheme.typography.labelLarge else MaterialTheme.typography.bodyMedium
    Row(Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 10.dp)) {
        Text(person, Modifier.weight(1f), style = style)
        Text(chr, Modifier.width(44.dp), style = style)
        Text(position, Modifier.width(96.dp), style = style)
        Text(genotype, Modifier.width(88.dp), style = style, color = if (muted) MaterialTheme.colorScheme.onSurfaceVariant else MaterialTheme.colorScheme.onSurface)
    }
}

/** A box in either tree: the pedigree convention is a square for men, a pill for women. */
@Composable
private fun PedigreeBox(sex: String, border: BorderStroke, onClick: (() -> Unit)? = null, content: @Composable () -> Unit) {
    val corner = when (Sex.of(sex)) {
        Sex.MALE -> 4.dp
        Sex.FEMALE -> NODE_H / 2
        Sex.UNKNOWN -> 8.dp
    }
    Surface(
        shape = RoundedCornerShape(corner),
        border = border,
        color = MaterialTheme.colorScheme.surface,
        modifier = Modifier.fillMaxSize().let { if (onClick != null) it.clickable(onClick = onClick) else it },
    ) {
        Box(contentAlignment = Alignment.Center) { content() }
    }
}

/**
 * The pedigree for one rsid (InheritanceTree.tsx): each person's genotype, and on every line the
 * allele that parent passed on. Risk alleles are red, ambiguous lines dashed, impossible ones red.
 */
@Composable
private fun InheritanceTree(persons: List<Person>, relationships: List<Relationship>, calls: List<FamilyCall>, entry: KbEntry?) {
    val t = LocalStrings.current
    val danger = MaterialTheme.colorScheme.error
    val warn = severityColor(4)
    val nodes = layoutPedigree(persons, relationships)
    val callOf = calls.associate { it.personId to it.call }
    val risk = entry?.riskAllele
    val edges = pedigreeEdges(nodes).map { e ->
        val child = callOf[e.to.person.id]
        val origins = if (child != null) alleleOrigins(child, e.to.parents.map { ParentCall(it, callOf[it]) }) else emptyList()
        val mine = origins.filter { it.from == e.from.person.id }.map { it.allele }
        val impossible = origins.firstOrNull()?.from == Origin.IMPOSSIBLE
        e.copy(
            color = if (impossible) danger else null,
            width = if (mine.isNotEmpty()) 2.dp else 1.2.dp,
            dashed = origins.isNotEmpty() && !impossible && mine.isEmpty(),
            label = mine.joinToString("").ifEmpty { null },
            labelColor = if (risk != null && risk in mine) danger else null,
        )
    }
    Pedigree(nodes, edges) { n ->
        val c = callOf[n.person.id]
        val copies = if (c != null && risk != null) listOf(c.a1, c.a2).count { it == risk } else 0
        val outline = when (copies) {
            2 -> danger
            1 -> warn
            else -> MaterialTheme.colorScheme.outlineVariant
        }
        PedigreeBox(n.person.sex, BorderStroke(if (copies > 0) 2.dp else 1.dp, outline)) {
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Text(n.person.displayName, fontWeight = FontWeight.SemiBold, maxLines = 1)
                if (c == null) {
                    Text(t("inheritanceTree.notTyped"), color = MaterialTheme.colorScheme.onSurfaceVariant, style = MaterialTheme.typography.bodySmall)
                } else {
                    Text(
                        buildAnnotatedString {
                            for (a in listOf(c.a1, c.a2)) {
                                if (a == risk) withStyle(SpanStyle(color = danger, fontWeight = FontWeight.Bold)) { append(a) } else append(a)
                            }
                        },
                        fontSize = 16.sp,
                    )
                }
            }
        }
    }
    Text(
        rich(
            t("inheritanceTree.legend") + (risk?.let { " " + t("inheritanceTree.riskLegend", "allele" to it) } ?: ""),
            mapOf("r" to SpanStyle(color = danger, fontWeight = FontWeight.Bold)),
        ),
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
}

/** Everyone as a pedigree (FamilyTree.tsx); a person with genotypes opens their report. */
@Composable
fun FamilyTree(persons: List<Person>, relationships: List<Relationship>, counts: Map<String, Int>, onOpen: (Person) -> Unit) {
    val t = LocalStrings.current
    val nodes = layoutPedigree(persons, relationships)
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Pedigree(nodes, pedigreeEdges(nodes), Modifier.padding(horizontal = 8.dp)) { n ->
            val p = n.person
            val snps = counts[p.id] ?: 0
            val border = if (snps > 0) BorderStroke(1.5.dp, MaterialTheme.colorScheme.primary) else BorderStroke(1.dp, MaterialTheme.colorScheme.outlineVariant)
            PedigreeBox(p.sex, border, onClick = if (snps > 0) ({ onOpen(p) }) else null) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(p.displayName, fontWeight = FontWeight.SemiBold, maxLines = 1)
                    Text(
                        listOfNotNull(
                            p.birthYear?.let { t("peoplePage.born", "year" to it) },
                            if (snps > 0) t("peoplePage.snps", "n" to snps.grouped()) else t("peoplePage.noGenotypes"),
                        ).joinToString(" · "),
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        textAlign = TextAlign.Center,
                        maxLines = 2,
                    )
                }
            }
        }
        if (persons.size > 1 && relationships.isEmpty()) {
            Text(t("peoplePage.treeHint"), Modifier.padding(horizontal = 16.dp), style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }
    }
}

File: mobile/ios/Hearth/Native/UI/FamilyScreen.swift
import SwiftUI

/// Family lookup (FamilyPage.tsx): find a marker by rsid, gene, drug or condition, then see every
/// person's genotype at it and, when parents are set, who inherited which allele from whom.
struct FamilyScreen: View {
    let repo: Repo
    let kb: Kb

    @State private var q = ""
    @State private var looked: String?
    @State private var rows: [FamilyCall]?
    @State private var persons: [Person] = []
    @State private var relationships: [Relationship] = []
    @State private var failure: String?

    private static let rsidPattern = try! NSRegularExpression(pattern: #"^rs\d+$"#, options: [.caseInsensitive])

    private var trimmed: String { q.trimmingCharacters(in: .whitespacesAndNewlines) }

    private var isRsid: Bool {
        let s = trimmed
        return FamilyScreen.rsidPattern.firstMatch(in: s, range: NSRange(location: 0, length: (s as NSString).length)) != nil
    }

    private var hits: [KbEntry] {
        isRsid ? [] : Array(KbLogic.search(kb, q).prefix(12))
    }

    private var entry: KbEntry? {
        guard let looked else { return nil }
        return kb.entries.first { $0.rsid == looked }
    }

    var body: some View {
        NavigationStack {
            List {
                if isRsid {
                    Section {
                        Button {
                            lookup(trimmed)
                        } label: {
                            Label(t("familyPage.lookUp") + " " + trimmed, systemImage: "magnifyingglass")
                        }
                    }
                }
                if !hits.isEmpty {
                    Section {
                        ForEach(hits) { e in
                            Button {
                                lookup(e.rsid)
                            } label: {
                                VStack(alignment: .leading, spacing: 2) {
                                    Text(e.rsid).font(.caption).foregroundStyle(.secondary)
                                    Text(verbatim: "\(e.gene) — \(e.name)").foregroundStyle(.primary)
                                    Text(e.summary).font(.footnote).foregroundStyle(.secondary).lineLimit(2)
                                }
                            }
                        }
                    }
                }
                if let found = rows, let looked {
                    result(looked, found)
                }
            }
            .navigationTitle(t("familyPage.title"))
            .searchable(text: $q, prompt: Text(t("familyPage.searchPlaceholder")))
            .textInputAutocapitalization(.never)
            .autocorrectionDisabled()
            .onSubmit(of: .search) { if isRsid { lookup(trimmed) } }
            .alert(failure ?? "", isPresented: Binding(get: { failure != nil }, set: { if !$0 { failure = nil } })) {
                Button(t("common.close"), role: .cancel) {}
            }
        }
        .onAppear(perform: load)
    }

    @ViewBuilder
    private func result(_ rsid: String, _ found: [FamilyCall]) -> some View {
        Section {
            VStack(alignment: .leading, spacing: 2) {
                Text(rsid).font(.title2.weight(.semibold))
                if let entry {
                    Text(verbatim: "\(entry.gene) — \(entry.name)").foregroundStyle(.secondary)
                }
            }
        }
        if !found.isEmpty && !relationships.isEmpty {
            Section {
                InheritanceTree(persons: persons, relationships: relationships, calls: found, entry: entry)
                    .listRowInsets(EdgeInsets(top: 8, leading: 0, bottom: 8, trailing: 0))
            } header: {
                Text(t("familyPage.whoInherited"))
            }
        }
        if found.isEmpty {
            Section {
                Text(t("familyPage.notGenotyped")).foregroundStyle(.secondary)
            }
        } else {
            Section {
                Grid(alignment: .leading, horizontalSpacing: 12, verticalSpacing: 10) {
                    GridRow {
                        Text(t("familyPage.person"))
                        Text(t("familyPage.chr"))
                        Text(t("familyPage.position"))
                        Text(t("familyPage.genotype"))
                    }
                    .font(.subheadline.weight(.semibold))
                    Divider()
                    ForEach(persons) { p in
                        let call = found.first { $0.personId == p.id }?.call
                        GridRow {
                            Text(p.displayName)
                            Text(call?.chromosome ?? "–")
                            Text(call.map { String($0.position) } ?? "–").monospacedDigit()
                            if let call {
                                Text(verbatim: "\(call.a1)/\(call.a2)").monospaced()
                            } else {
                                Text(t("familyPage.notOnChip")).foregroundStyle(.secondary)
                            }
                        }
                    }
                }
            }
        }
    }

    // MARK: - Actions

    /// People and links are re-read whenever the tab appears (they change on the People tab), and
    /// the marker on screen is looked up again against them.
    private func load() {
        do {
            persons = try repo.persons()
            relationships = try repo.relationships()
            if let looked { rows = try repo.familyAt(looked) }
        } catch {
            failure = t("native.error", ["error": error.localizedDescription])
        }
    }

    private func lookup(_ rsid: String) {
        q = rsid
        looked = rsid
        do {
            rows = try repo.familyAt(rsid)
        } catch {
            failure = t("native.error", ["error": error.localizedDescription])
        }
    }
}

/// The pedigree for one rsid (InheritanceTree.tsx): each person's genotype, and on every line the
/// allele that parent passed on. Risk alleles are red, ambiguous lines dashed, impossible ones red.
struct InheritanceTree: View {
    let persons: [Person]
    let relationships: [Relationship]
    let calls: [FamilyCall]
    let entry: KbEntry?

    private var danger: Color { severityColor(7) }
    private var warn: Color { severityColor(4) }

    var body: some View {
        let nodes = Inheritance.layoutPedigree(persons, relationships)
        let callOf = Dictionary(calls.map { ($0.personId, $0.call) }, uniquingKeysWith: { _, last in last })
        let risk = entry?.riskAllele
        let edges = styledEdges(nodes, callOf, risk)
        VStack(alignment: .leading, spacing: 8) {
            Pedigree(nodes: nodes, edges: edges) { n in
                let c = callOf[n.person.id]
                let copies = c.map { call in risk.map { r in [call.a1, call.a2].filter { $0 == r }.count } ?? 0 } ?? 0
                PedigreeBox(
                    sex: n.person.sex,
                    border: copies == 2 ? danger : (copies == 1 ? warn : Color(uiColor: .separator)),
                    lineWidth: copies > 0 ? 2 : 1
                ) {
                    VStack(spacing: 2) {
                        Text(n.person.displayName).font(.subheadline.weight(.semibold)).lineLimit(1)
                        if let c {
                            Text(genotype(c, risk)).font(.body.monospaced())
                        } else {
                            Text(t("inheritanceTree.notTyped")).font(.caption).foregroundStyle(.secondary)
                        }
                    }
                }
            }
            Text(rich(legend(risk), colors: ["r": danger]))
                .font(.footnote)
                .foregroundStyle(.secondary)
                .padding(.horizontal)
        }
    }

    /// Each line with the allele that parent passed on: bold when it names one, dashed when either
    /// parent could have, red when no assignment works, the letter red when it is the risk allele.
    private func styledEdges(_ nodes: [PedigreeNode], _ callOf: [String: Call], _ risk: String?) -> [PedigreeEdge] {
        PedigreeEdge.of(nodes).map { e in
            var styled = e
            let origins = callOf[e.to.person.id].map { child in
                Inheritance.alleleOrigins(child, e.to.parents.map { Inheritance.ParentCall(id: $0, call: callOf[$0]) })
            } ?? []
            let mine = origins.filter { $0.from == e.from.person.id }.map(\.allele)
            let impossible = origins.first?.from == Inheritance.Origin.impossible
            styled.color = impossible ? danger : nil
            styled.width = mine.isEmpty ? 1.2 : 2
            styled.dashed = !origins.isEmpty && !impossible && mine.isEmpty
            styled.label = mine.isEmpty ? nil : mine.joined()
            if let risk, mine.contains(risk) { styled.labelColor = danger }
            return styled
        }
    }

    /// The two alleles, a risk allele bold and red.
    private func genotype(_ c: Call, _ risk: String?) -> AttributedString {
        var out = AttributedString()
        for a in [c.a1, c.a2] {
            var part = AttributedString(a)
            if a == risk {
                part.foregroundColor = danger
                part.inlinePresentationIntent = .stronglyEmphasized
            }
            out += part
        }
        return out
    }

    private func legend(_ risk: String?) -> String {
        var text = t("inheritanceTree.legend")
        if let risk { text += " " + t("inheritanceTree.riskLegend", ["allele": risk]) }
        return text
    }
}
